src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVMKernels.cpp

```diff
@@ -663,6 +663,356 @@ INST_FluxKernel( 5, FaceElementStencil );
+        for( localIndex k = 0; k < 3; ++k )
+        {
+          sum += dPhaseRelPerm_dPhaseVolFrac[k][j] / phaseVisc[k]
+                 * phaseOutflux[j];
+        }
+        f[i][j] *= sum;
+      }
+    }
+    phaseCFLNumber = f[0][0] + f[1][1];
+    phaseCFLNumber += sqrt( phaseCFLNumber*phaseCFLNumber - 4 * ( f[0][0]*f[1][1] - f[1][0]*f[0][1] ) );
+    phaseCFLNumber = 0.5 * fabs( phaseCFLNumber ) / poreVol;
+  }
+}
+
+
+template< localIndex NC >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void
+CFLKernel::
+  computeCompCFL( real64 const & poreVol,
+                  arraySlice1d< real64 const > compDens,
+                  arraySlice1d< real64 const > compOutflux,
+                  real64 & compCFLNumber )
+{
+
+
+  compCFLNumber = 0.0;
+  for( localIndex ic = 0; ic < NC; ++ic )
+  {
+    real64 const compMoles = compDens[ic] * poreVol;
+    if( compMoles > 1e-12 )
+    {
+      real64 const CFL = compOutflux[ic] / compMoles;
+      if( CFL > compCFLNumber )
+      {
+        compCFLNumber = CFL;
+      }
+    }
+  }
+}
+
+template< localIndex NC, localIndex NP >
+void
+CFLKernel::
+  launch( localIndex const size,
+          arrayView1d< real64 const > const & volume,
+          arrayView1d< real64 const > const & porosityRef,
+          arrayView2d< real64 const > const & pvMult,
+          arrayView2d< real64 const > const & compDens,
+          arrayView3d< real64 const > const & phaseRelPerm,
+          arrayView4d< real64 const > const & dPhaseRelPerm_dPhaseVolFrac,
+          arrayView3d< real64 const > const & phaseVisc,
+          arrayView2d< real64 const > const & phaseOutflux,
+          arrayView2d< real64 const > const & compOutflux,
+          real64 & maxPhaseCFLNumber,
+          real64 & maxCompCFLNumber )
+{
+  RAJA::ReduceMax< parallelDeviceReduce, real64 > regionPhaseCFLNumber( 0.0 );
+  RAJA::ReduceMax< parallelDeviceReduce, real64 > regionCompCFLNumber( 0.0 );
+
+  forAll< parallelDevicePolicy<> >( size, [=] GEOSX_HOST_DEVICE ( localIndex const ei )
+  {
+    real64 const poreVol = volume[ei] * porosityRef[ei] * pvMult[ei][0];
+
+    // phase CFL number
+    real64 cellPhaseCFLNumber = 0.0;
+    computePhaseCFL< NP >( poreVol,
+                           phaseRelPerm[ei][0],
+                           dPhaseRelPerm_dPhaseVolFrac[ei][0],
+                           phaseVisc[ei][0],
+                           phaseOutflux[ei],
+                           cellPhaseCFLNumber );
+    regionPhaseCFLNumber.max( cellPhaseCFLNumber );
+
+    // component CFL number
+    real64 cellCompCFLNumber = 0.0;
+    computeCompCFL< NC >( poreVol,
+                          compDens[ei],
+                          compOutflux[ei],
+                          cellCompCFLNumber );
+    regionCompCFLNumber.max( cellCompCFLNumber );
+
+  } );
+
+  maxPhaseCFLNumber = LvArray::math::max( maxPhaseCFLNumber, regionPhaseCFLNumber.get() );
+  maxCompCFLNumber = LvArray::math::max( maxCompCFLNumber, regionCompCFLNumber.get() );
```

Comment:
I'm not a big fan of in-out parameters, they are too easy to misuse (accidentally provide uninitialized value, expecting the function to overwrite it, and get garbage out). Maybe you could just return the subregion max CFL via `maxPhaseCFLNumber` and `maxCompCFLNumber` and have the calling code do the max over (sub)regions. It's a little more verbose (two extra local variables in the calling code), but the code is easier to reason about IMO, since looping over subregions and taking the max are both in one place.

Reply:
Now I just return the subRegion max CFL and then the calling code takes the max over the subRegions. I made the same change in `calculateResidualNorm` for both `CompositionalMultiphaseFVM` and `CompositionalMultiphaseHybridFVM` to avoid the bug that I had introduced a few month ago by wrongly setting `localResidualNorm` to 0 for each subRegion.